pkg/bspatch/bspatch.go

```diff
@@ -59,34 +67,51 @@ func Reader(oldbin io.Reader, newbin io.Writer, patchf io.Reader) error {
 
 // File applies a BSDIFF4 patch (using oldfile and patchfile) to create the newfile
 func File(oldfile, newfile, patchfile string) error {
-	oldbs, err := ioutil.ReadFile(oldfile)
+	oldf, err := os.Open(oldfile)
+	if err != nil {
+		return fmt.Errorf("could not open oldfile '%v': %v", oldfile, err.Error())
+	}
+	defer oldf.Close()
+
+	newf, err := os.Create(newfile)
 	if err != nil {
-		return fmt.Errorf("could not read oldfile '%v': %v", oldfile, err.Error())
+		return fmt.Errorf("could not open or create newfile '%v': %v", newfile, err.Error())
 	}
+
 	patchbs, err := ioutil.ReadFile(patchfile)
 	if err != nil {
 		return fmt.Errorf("could not read patchfile '%v': %v", patchfile, err.Error())
 	}
-	newbytes, err := patchb(oldbs, patchbs)
+
+	newfw := bufio.NewWriterSize(newf, WriteBufferSize)
+	err = patchStream(oldf, newfw, patchbs)
+
 	if err != nil {
+		newf.Close()
+		os.Remove(newfile)
 		return fmt.Errorf("bspatch: %v", err.Error())
 	}
-	if err := ioutil.WriteFile(newfile, newbytes, 0644); err != nil {
-		return fmt.Errorf("could not create newfile '%v': %v", newfile, err.Error())
-	}
+
+	newfw.Flush()
+	newf.Close()
 	return nil
 }
 
-func patchb(oldfile, patch []byte) ([]byte, error) {
-	oldsize := len(oldfile)
-	var newsize int
-	header := make([]byte, 32)
-	buf := make([]byte, 8)
-	var lenread int
-	var i int
-	ctrl := make([]int, 3)
+func patchStream(oldf io.ReadSeeker, newf io.Writer, patch []byte) error {
+	HeaderLen := 32
 
-	f := bytes.NewReader(patch)
+	// Reused container vars
+	var lenread int64
+	var errmsg string
+	header := make([]byte, HeaderLen)
```

Comment:
You can just allocate `var header [HeaderLen]byte` and turn it into a slice via `&header`.